### ISSUE TITLE
Is figure of Batch normalization correct?

### ISSUE THREAD

**Issue by Jingnan-Jia** [titu1994/Keras-Group-Normalization#8]
![original figure of BN](https://raw.githubusercontent.com/titu1994/Keras-Group-Normalization/master/images/normalization_techniques.PNG)
As far as I am concerned, the representation of batch normalization is not correct in the original paper. I post the issue here for discussion.
I think the batch normalization should be like the following figure.
![BN](https://user-images.githubusercontent.com/34941987/72001195-331ea500-3245-11ea-8e55-171e5da8edff.png)

The key point is how to calculate mean and std.
With feature maps' shape as `(batch_size, channel_number, width, height)`,
`mean = X.mean(axis=(0, 2, 3), keepdims=True)`
or
`mean = X.mean(axis=(0, 1), keepdims=True)`

Which one is correct?

**Comment by titu1994** [comment]
Batch norm is calculated on axis (0, 2, 3), and the paper illustration demonstrates this axis properly. 

Your image demonstrates normalization over axis (0, 1). As far as I know, that will Normalize the pixelwise representation over batch and channel, which if you think about it, is  equivalent to taking a bunch of RGB images, and squashing the RGB into a singular channel (say grayscale) and take mean and STD over all the images in that batch.

**Comment by Jingnan-Jia** [comment]
Thanks for your respons. Please see the [paper](https://arxiv.org/pdf/1502.03167.pdf) in which batch normalization was first proposed.  

![image](https://user-images.githubusercontent.com/34941987/72005038-9cbfa300-3287-11ea-8c16-36e6f5c28baa.png)

From the equation we can find that every dimention have different means and variances. So for a image with dimention 128\*128, it should have 128\*128 means and variance, instead just one. Isnot it?

**Comment by Jingnan-Jia** [comment]
![image](https://user-images.githubusercontent.com/34941987/72006024-ca0d5080-3289-11ea-97f8-196e1dceffa5.png)

I am confused by the equation. Is \[ \mu \] a matrics? or vector? or scalar? (Because the author did not sum all elements in x)

**Comment by titu1994** [comment]
Mu is a vector is length number of input channels (say C). For broadcasting, it is left as a tensor of shape (1, C, 1, 1) with keepdims. 

For tensorflow, we generally set axis = -1, and therefore refer to this line in the reduction axis of Tensorflow Batch Norm computation - https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/keras/layers/normalization.py#L669.

Beta and gamma are also vectors which compute the moving average statistics of Mu and Sigma.

**Comment by titu1994** [comment]
> Thanks for your respons. Please see the [paper](https://arxiv.org/pdf/1502.03167.pdf) in which batch normalization was first proposed.
> 
> ![image](https://user-images.githubusercontent.com/34941987/72005038-9cbfa300-3287-11ea-8c16-36e6f5c28baa.png)
> 
> From the equation we can find that every dimention have different means and variances. So for a image with dimention 128_128, it should have 128_128 means and variance, instead just one. Isnot it?

No. An image is (in RGB sense), a tensor of shape (1, H, W, 3). The 1 is a batch size of 1. Batch Norm will compute a norm of shape (1, 1, 1, 3), over the RGB channels. That's all. You can flatten this tensor into a vector of length 3, if you don't want to bother with broadcasting. 

This RGB can be represented by C as the number of input channels.

**Comment by Jingnan-Jia** [comment]
In this [post](https://mlexplained.com/2018/11/30/an-overview-of-normalization-methods-in-deep-learning/), you can see the following sentence.

> If the batch size is 1, the variance is 0 so batch normalization cannot be applied. 

Why? If the batch normalization is what you said, why batch_size=1 will lead to division by zero?

**Comment by titu1994** [comment]
How do you compute mean and std of a single sample? Mean is still ok as you divide X by 1.

But Standard deviation of a single sample is 0. Therefore sigma is 0. Division by sigma itself will give you NaN.

**Comment by Jingnan-Jia** [comment]
But in 3D medical image segmentation area, nearly all the researchers are using batch size = 1 (because of GPU limitation) with Batch Norm, when they use 3D U-Net. Ant get good results without errors.

**Comment by titu1994** [comment]
I will be closing this issue as it does not pertain to the repository or it's code. Feel free to continue the conversation on this thread.

**Comment by titu1994** [comment]
> But in 3D medical image segmentation area, nearly all the researchers are using batch size = 1 (because of GPU limitation) with Batch Norm, when they use 3D U-Net. Ant get good results without errors.

Look at the equation in algorithm 1, it has a sigma + epsilon inside the root. That epsilon is generally a small number like 1e-5 or 1e-3. So even if std is 0, the sqrt of just that Epsilon value is not NaN. So it avoids crashing the program.

**Comment by Jingnan-Jia** [comment]
Thanks for your patience. Okay, from the figure,  I wonder if batch_size=1, would batch normalization be degraded to instance normalization? It yes, then the reason why batch_size=1 can work may be not because that epsilon can be obtained by Epsilon=1e-5, but because epsilon can be obtained by the H*W pixels.

**Comment by Jingnan-Jia** [comment]
> > Thanks for your respons. Please see the [paper](https://arxiv.org/pdf/1502.03167.pdf) in which batch normalization was first proposed.
> > ![image](https://user-images.githubusercontent.com/34941987/72005038-9cbfa300-3287-11ea-8c16-36e6f5c28baa.png)
> > From the equation we can find that every dimention have different means and variances. So for a image with dimention 128_128, it should have 128_128 means and variance, instead just one. Isnot it?
> 
> No. An image is (in RGB sense), a tensor of shape (1, H, W, 3). The 1 is a batch size of 1. Batch Norm will compute a norm of shape (1, 1, 1, 3), over the RGB channels. That's all. You can flatten this tensor into a vector of length 3, if you don't want to bother with broadcasting.
> 
> This RGB can be represented by C as the number of input channels.

Do you mean that in the equation, k respresent the index of channels  instead of the index of pixels ？

**Comment by Jingnan-Jia** [comment]
> How do you compute mean and std of a single sample? Mean is still ok as you divide X by 1.
> 
> But Standard deviation of a single sample is 0. Therefore sigma is 0. Division by sigma itself will give you NaN.

I mean I can compute mean and std just of a single sample because the sample consists of `H*W` pixels. I can get the mean and std using the `H*W` pixels.

**Comment by titu1994** [comment]
> Thanks for your patience. Okay, from the figure, I wonder if batch_size=1, would batch normalization be degraded to instance normalization? It yes, then the reason why batch_size=1 can work may be not because that epsilon can be obtained by Epsilon=1e-5, but because epsilon can be obtained by the H*W pixels.

The equation of std dev = sqrt (1/n sum (xi - m)^2) right ? At N=1, m = xi. So sqrt of 0 basically. Doesn't matter how many dimensions X or m is.

**Comment by titu1994** [comment]
> > > Thanks for your respons. Please see the [paper](https://arxiv.org/pdf/1502.03167.pdf) in which batch normalization was first proposed.
> > > ![image](https://user-images.githubusercontent.com/34941987/72005038-9cbfa300-3287-11ea-8c16-36e6f5c28baa.png)
> > > From the equation we can find that every dimention have different means and variances. So for a image with dimention 128_128, it should have 128_128 means and variance, instead just one. Isnot it?
> > 
> > 
> > No. An image is (in RGB sense), a tensor of shape (1, H, W, 3). The 1 is a batch size of 1. Batch Norm will compute a norm of shape (1, 1, 1, 3), over the RGB channels. That's all. You can flatten this tensor into a vector of length 3, if you don't want to bother with broadcasting.
> > This RGB can be represented by C as the number of input channels.
> 
> Do you mean that in the equation, k respresent the index of channels instead of the index of pixels ？

Yes.

**Comment by Jingnan-Jia** [comment]
> > Thanks for your patience. Okay, from the figure, I wonder if batch_size=1, would batch normalization be degraded to instance normalization? It yes, then the reason why batch_size=1 can work may be not because that epsilon can be obtained by Epsilon=1e-5, but because epsilon can be obtained by the H*W pixels.
> 
> The equation of std dev = sqrt (1/n sum (xi - m)^2) right ? At N=1, m = xi. So sqrt of 0 basically. Doesn't matter how many dimensions X or m is. They are not equal.

The key point is that you said 

> std dev = sqrt (1/n sum (xi - m)^2), At N=1, m = xi. So sqrt of 0 basically

But in my mind, at N (batch size)=1, m=sum(xi, axis=(0, 2, 3))/ (batch_size\*H\*W). m is a scalar but xi is a matricx with shape (1, 1, H, W). so m is not equal to xi. So std dev can be obtained normally.

Is something wrong? 

---------------------------------------------------------------------------------------------------------
Let me try explain how to calculate std dev with batch_size=1.

lets set 
`batch_size=10`,    
`X.shape=(batch_size=10, channel_number=32, H=128, W=128)`,    
`X_channel_1.shape=(batch_size=10, channel_number=1, H=128, W=128)`,   
According to the BN figure, at first, lets calculate the mean and std_dev in the first channel.

`mean_channel_1 = sum(X_channel_1, axis=(0,2,3)) / (10*128*128)`,  

`mean_channel_1` is a scalar now. Right?

`std_dev_channel_1 = sqrt (variance) = sqrt( sum( (X_channel_1 - mean_channel_1 )^2, axis=(0,2,3)) / (10*128*128) )`  

Here,   
`X_channel_1.shape=(10,1,128,128)`, 
`mean_channel_1` is a scalar, 
`(X_channel_1 - mean_channel_1 )^2.shape=(10,1,128,128)`, `sum( (X_channel_1 - mean_channel_1 )^2, axis=(0,2,3))` is a scalar, 
so `std_dev` is a scalar. Right?

`bn_channel_1 = (X_channel_1 - mean_of_channel_1) / std_dev_channel_1`,  
`bn_channel_1.shape = (10,1,128,128)`. Right?

Then calculate `bn_channel_2, 3, ..., 32`. Finally we get `X_bn.shape= (10,32,128,128)`. Right?

If all the above processes are right, then I can show how to calculate `mean` and `std_dev` with `batch_size=1`:
Lets set:   
`batch_size=1`,    
`X.shape=(batch_size=1, channel_number=32, H=128, W=128)`,    
`X_channel_1.shape=(batch_size=1, channel_number=1, H=128, W=128)`,   
According to the BN figure, at first, lets calculate the mean and std_dev in the first channel.

`mean_channel_1 = sum(X_channel_1, axis=(0,2,3)) / (1*128*128)`   
`mean_channel_1` is a scalar now. Right?    

`std_dev_channel_1 = sqrt (variance) = sqrt( sum( (X_channel_1 - mean_channel_1 )^2, axis=(0,2,3)) / (1*128*128) )`,    
Here, 
`X_channel_1.shape=(1,1,128,128)`, 
 `mean_channel_1` is a scalar, `(X_channel_1 - mean_channel_1 )^2.shape=(1,1,128,128)`, 
 `sum( (X_channel_1 - mean_channel_1 )^2, axis=(0,2,3))` is a scalar, 
so `std_dev` is a scalar. Right?

`bn_channel_1 = (X_channel_1 - mean_of_channel_1) / std_dev_channel_1`,  
`bn_channel_1.shape = (1,1,128,128)`. Right?

Then calculate bn_channel_2, 3, ..., 32. Finally we get `X_bn.shape= (1,32,128,128)`. Right?

The key point is that you said 

> std dev = sqrt (1/n sum (xi - m)^2), At N=1, m = xi. So sqrt of 0 basically

But in my mind, at N (batch size)=1, m=sum(xi, axis=(0, 2, 3))/ (batch_size\*H\*W).

Is something wrong? Looking forward to your reply. 

Sincerely.

**Comment by Jingnan-Jia** [comment]
Please see this [answer](https://stackoverflow.com/questions/59651396/how-to-calculate-batch-normalization-with-python/59654494?noredirect=1#comment105470721_59654494)

> Those people, who think that batch_size=1 will lead to division by zero assume that std is calculated separately for each pixel. But even for batch_size=1 it is calculated across H*W pixels.
